### PR TITLE
feat: Provisioner Plugin POC

### PR DESCRIPTION
**Pull request by jvmakine**
This adds a Cloudformation based Plugin for provisioning databases. Some things to note:
- This is a POC and not intended for production
- Still missing Plan API
- We are passing requited underlying subnet group id as a resource property of a dependent "FTL Cluster" resource. This needs to be eventually populated by FTL bootstrap process, when creating the cluster.
- `dev` runner is just a helper to try this out locally. It needs to evolve something much more generic and useful in the future.

This creates a single Cloudformation Stack / module, with all module resources in it

Screenshot 2024-09-18 at 11 38 02 AM

### CODE REVIEW

File: cmd/provisioners/ftl-provisioner-cloudformation/cfutil/cfutil.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+		}
+		if desc.Status == types.ChangeSetStatusFailed {
+			// Unfortunately, there does not seem to be a better way to do this
+			if *desc.StatusReason == "The submitted information didn't contain changes. Submit different information to create a change set." {
+				// clean up the changeset if there were no changes
+				_, err := client.DeleteChangeSet(ctx, &cloudformation.DeleteChangeSetInput{
+					ChangeSetName: &changeSet,
+					StackName:     &stack,
+				})
+				return false, fmt.Errorf("failed to delete change-set: %w", err)
```
Comment (jvmakine) [inline cmd/provisioners/ftl-provisioner-cloudformation/cfutil/cfutil.go:78]: Noticed a bug here if there are no changes, will fix this in a followup PR.

File: cmd/provisioners/ftl-provisioner-cloudformation/main.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+	"context"
+
+	"connectrpc.com/connect"
+
+	ftlv1 "github.com/TBD54566975/ftl/backend/protos/xyz/block/ftl/v1"
+	"github.com/TBD54566975/ftl/backend/protos/xyz/block/ftl/v1beta1/provisioner/provisionerconnect"
+	"github.com/TBD54566975/ftl/common/plugin"
+)
+
+type CloudformationProvisioner struct{}
```
Comment (alecthomas) [inline cmd/provisioners/ftl-provisioner-cloudformation/main.go:13]: I'd move these into provisioner.go, it's weird to disconnect them like this.

File: cmd/provisioners/ftl-provisioner-cloudformation/provision.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+			ResourceID:   resource.ResourceId,
+			PropertyName: "db:endpoint-write",
+		})
+		cfutil.AddOutput(template.Outputs, goformation.GetAtt(clusterID, "ReadEndpoint.Address"), &cfutil.CloudformationOutputKey{
+			ResourceID:   resource.ResourceId,
+			PropertyName: "db:endpoint-read",
+		})
+		return nil
+	}
+	return errors.New("unsupported resource type")
```
Comment (alecthomas) [inline cmd/provisioners/ftl-provisioner-cloudformation/provision.go:116]:
```suggestion
	return fmt.Errorf("unsupported resource type %T", resource.Resource)
```

File: frontend/console/src/protos/xyz/block/ftl/v1/provisioner/plugin_connect.ts
Comment (alecthomas) [inline frontend/console/src/protos/xyz/block/ftl/v1/provisioner/plugin_connect.ts:1]: We should come up with some mechanism to exclude TS codegen for some proto files.

**Review by alecthomas — APPROVED** [review]
Noice.

File: backend/protos/xyz/block/ftl/v1beta1/provisioner/plugin.proto [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+message ProvisionRequest {
+  string ftl_cluster_id = 1;
+  string module = 2;
+
+  // The resource FTL thinks exists currently
+  repeated Resource existing_resources = 3;
+
+  // The resource FTL would like to exist after this provisioning run
+  repeated Resource new_resources = 4;
```
[the remark this answers is not in the document]
Comment (jvmakine) [reply backend/protos/xyz/block/ftl/v1beta1/provisioner/plugin.proto:48]: The idea is that the provisioner needs to calculate that from the old and new resources, if it needs to know what has changed. I think this is only needed if the provisioner uses an imperative approach to provisioning

File: cmd/provisioners/ftl-provisioner-cloudformation/cfutil/cfutil.go [hunk 5]
```diff
@@ -0,0 +1,164 @@
+package cfutil
```
[the remark this answers is not in the document]
Comment (jvmakine) [reply cmd/provisioners/ftl-provisioner-cloudformation/cfutil/cfutil.go:1]: 👍  I might just move these to the `main` package with rest of the provisioner. Separate package here didn't make as much sense as I thought.

File: backend/protos/xyz/block/ftl/v1beta1/provisioner/plugin.proto (hunk 4 above)
[the remark this answers is not in the document]
Comment (jvmakine) [reply backend/protos/xyz/block/ftl/v1beta1/provisioner/plugin.proto:48]: Ah, I see the confusion. This should be called `desired_resources` or something. Will rename

File: backend/protos/xyz/block/ftl/v1beta1/provisioner/plugin.proto [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  enum ProvisioningStatus {
+    UNKNOWN = 0;
+    RUNNING = 1;
+    SUCCEEDED = 2;
+    FAILED = 3;
+  }
+
+  ProvisioningStatus status = 1;
+  string error_message = 2;
+  repeated ResourceProperty properties = 3;
```
[the remark this answers is not in the document]
Comment (jvmakine) [reply backend/protos/xyz/block/ftl/v1beta1/provisioner/plugin.proto:76]: Those are the implementation specific properties populated during resource creation. I think I will change this to be more explicit on returning new Resource statuses as a follow up, though.

Idea is to return things like database urls, or something resource specific like subnet groups that a dependant resource might need, but the provisioner itself does not need to know about.

File: cmd/provisioners/ftl-provisioner-cloudformation/cfutil/cfutil.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (alecthomas) [reply cmd/provisioners/ftl-provisioner-cloudformation/cfutil/cfutil.go]: I did this in #2718

**Pull request merged**